crates/oxc_linter/src/rules/jest/prefer_hooks_in_order.rs

```diff
@@ -0,0 +1,1279 @@
+    ///         });
+    ///         it('throws an error', () => {
+    ///             // ...
+    ///         });
+    ///         afterEach(() => {
+    ///             clearLogger();
+    ///         });
+    ///         beforeEach(() => {
+    ///             mockLogger();
+    ///         });
+    ///         it('logs a message', () => {
+    ///             // ...
+    ///         });
+    ///     });
+    ///     afterAll(() => {
+    ///         removeMyDatabase();
+    ///     });
+    /// });
+    /// ```
+    ///
+    /// ```javascript
+    /// // valid
+    /// describe('foo', () => {
+    ///     beforeAll(() => {
+    ///         createMyDatabase();
+    ///     });
+    ///
+    ///     beforeEach(() => {
+    ///         seedMyDatabase();
+    ///     });
+    ///
+    ///     it('accepts this input', () => {
+    ///         // ...
+    ///     });
+    ///     it('returns that value', () => {
+    ///         // ...
+    ///     });
+    ///     describe('when the database has specific values', () => {
+    ///         const specificValue = '...';
+    ///         beforeEach(() => {
+    ///             seedMyDatabase(specificValue);
+    ///         });
+    ///         it('accepts that input', () => {
+    ///             // ...
+    ///         });
+    ///         it('throws an error', () => {
+    ///             // ...
+    ///         });
+    ///         beforeEach(() => {
+    ///             mockLogger();
+    ///         });
+    ///         afterEach(() => {
+    ///             clearLogger();
+    ///         });
+    ///         it('logs a message', () => {
+    ///             // ...
+    ///         });
+    ///     });
+    ///     afterAll(() => {
+    ///         removeMyDatabase();
+    ///     });
+    /// });
+    /// ```
+    ///
+    ///
+    /// This rule is compatible with [eslint-plugin-vitest](https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/prefer_hooks_in_order.md),
+    /// to use it, add the following configuration to your `.eslintrc.json`:
+    ///
+    /// ```json
+    /// {
+    ///   "rules": {
+    ///      "vitest/prefer_hooks_in_order": "error"
+    ///   }
+    /// }
+    PreferHooksInOrder,
+    style,
+);
+
+impl Rule for PreferHooksInOrder {
+    fn run_once(&self, ctx: &LintContext) {
+        let mut previous_hooks_map: FxHashMap<ScopeId, usize> = FxHashMap::default();
+        let mut node_collections: Vec<AstNode> = vec![];
+
+        for node in ctx.nodes().iter() {
+            if let AstKind::CallExpression(call_expr) = node.kind() {
+                let possible_jest_node = &PossibleJestNode { node, original: None };
+                Self::check(
+                    &mut previous_hooks_map,
+                    &mut node_collections,
+                    possible_jest_node,
+                    call_expr,
+                    ctx,
+                );
+            };
+        }
+    }
+}
+
+impl PreferHooksInOrder {
+    fn check<'a>(
```

Comment:
Seems we can be more simple:

1. Group `nodes` by `scope_id`
2. In each group, check if it's ordered, in a group, we only need a record a `previousHookIndex` and `currentHookIndex` like jest does.

Do you have interest to try it out?

Reply:
Updated.